### PR TITLE
[X86][APX] Convert store(cmov(load(x), y), x) to cstore(y, x)

### PR DESCRIPTION
**Pull request by phoebewang**

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -2927,6 +2927,24 @@ static X86::CondCode TranslateIntegerX86CC(ISD::CondCode SetCCOpcode) {
   }
 }
 
+static X86::CondCode getInvertedX86CC(X86::CondCode CC) {
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:2930]: We already have this in X86InstroInfo as X86::GetOppositeBranchCondition - can we share a helper in X86BaseInfo.h?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -52786,6 +52804,42 @@ static SDValue combineStore(SDNode *N, SelectionDAG &DAG,
...
+    if (StoredVal.getOpcode() == X86ISD::CMOV)
+      Cmov = StoredVal;
+    else if (StoredVal.getOpcode() == ISD::TRUNCATE &&
+             StoredVal.getOperand(0).getOpcode() == X86ISD::CMOV)
+      Cmov = StoredVal.getOperand(0);
+    else
+      return SDValue();
+
+    auto *Ld = dyn_cast<LoadSDNode>(St->getChain());
+    if (!Ld || !Ld->isSimple() || Ld->getBasePtr() != St->getBasePtr())
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:52820]: check for ext load?

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: Great, thank! I had a quick search but didn't find it 🤦‍♀️

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 2 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:52820]: We have both ext load for i16 and normal load for i32/i64. I think we don't need to check ext load special.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 3]
```diff
@@ -52786,6 +52786,43 @@ static SDValue combineStore(SDNode *N, SelectionDAG &DAG,
...
+      return SDValue();
+
+    bool InvertCC = false;
+    SDValue V = SDValue(Ld, 0);
+    if (V == Cmov.getOperand(1))
+      InvertCC = true;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:52808]: Tested?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 4]
```diff
@@ -52786,6 +52786,43 @@ static SDValue combineStore(SDNode *N, SelectionDAG &DAG,
...
+    else if (V != Cmov.getOperand(0))
+      return SDValue();
+
+    SDVTList Tys = DAG.getVTList(MVT::Other);
+    SDValue CC = Cmov.getOperand(2);
+    SDValue Src = DAG.getAnyExtOrTrunc(Cmov.getOperand(!InvertCC), dl, VT);
+    if (InvertCC)
+      CC = DAG.getTargetConstant(
+          GetOppositeBranchCondition(
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:52817]: Is it a good time to rename it? Probably `getOppositeCondCode`

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 5]
```diff
@@ -52786,6 +52786,43 @@ static SDValue combineStore(SDNode *N, SelectionDAG &DAG,
...
+    if (StoredVal.getOpcode() == X86ISD::CMOV)
+      Cmov = StoredVal;
+    else if (StoredVal.getOpcode() == ISD::TRUNCATE &&
+             StoredVal.getOperand(0).getOpcode() == X86ISD::CMOV)
+      Cmov = StoredVal.getOperand(0);
+    else
+      return SDValue();
+
+    auto *Ld = dyn_cast<LoadSDNode>(St->getChain());
+    if (!Ld || !Ld->isSimple() || Ld->getBasePtr() != St->getBasePtr())
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:52802]: Add negative test for this?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 6]
```diff
@@ -52786,6 +52786,43 @@ static SDValue combineStore(SDNode *N, SelectionDAG &DAG,
     }
   }
 
+  // Convert store(cmov(load(x), y), x) to cstore(y, x).
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:52789]: Maybe write out the CC operand in the comment?

Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:52789]: Done.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 5 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:52802]: Done.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 3 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:52808]: Covered by `cfcmov16mr`.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 4 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:52817]: I found it's a common interface used by other targets too. Better not to change it.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 7]
```diff
@@ -52786,6 +52786,44 @@ static SDValue combineStore(SDNode *N, SelectionDAG &DAG,
...
+    if (StoredVal.getOpcode() == X86ISD::CMOV)
+      Cmov = StoredVal;
+    else if (StoredVal.getOpcode() == ISD::TRUNCATE &&
+             StoredVal.getOperand(0).getOpcode() == X86ISD::CMOV)
+      Cmov = StoredVal.getOperand(0);
+    else
+      return SDValue();
+
+    auto *Ld = dyn_cast<LoadSDNode>(St->getChain());
+    if (!Ld || !Ld->isSimple() || Ld->getBasePtr() != St->getBasePtr())
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:52803]: Better to have test for `Ld->getBasePtr() != St->getBasePtr()` too.

**Review by KanRobert — APPROVED** [review]
LGTM with one suggestion

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 7 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:52803]: Done.

**Pull request merged**